orderer/consensus/hlmirbft/chain.go

```diff
@@ -1056,146 +1091,224 @@ func (c *Chain) triggerCatchup(sn *raftpb.Snapshot) {
 	case <-c.doneC:
 	}
 }
+
 //JIRA FLY2-48 proposed changes: fetch request from request store
-func (c *Chain) fetchRequest(ack *msgs.RequestAck) (*msgs.Request,error){
+func (c *Chain) fetchRequest(ack *msgs.RequestAck) (*msgs.Request, error) {
 
 	reqByte, err := c.Node.ReqStore.GetRequest(ack)
 	if err != nil {
 		return nil, errors.WithMessage(err, "Cannot Fetch Request")
 	}
 	if reqByte == nil {
-		return nil,errors.Errorf("reqstore should have request if we are committing it")
+		return nil, errors.Errorf("reqstore should have request if we are committing it")
 	}
 	reqMsg := &msgs.Request{}
-	err = proto.Unmarshal(reqByte,reqMsg)
+	err = proto.Unmarshal(reqByte, reqMsg)
 	if err != nil {
-		return nil,errors.WithMessage(err, "Cannot Unmarshal Request")
+		return nil, errors.WithMessage(err, "Cannot Unmarshal Request")
 	}
-	return reqMsg,nil
+	return reqMsg, nil
 }
+
 //FLY2-48 proposed changes
 // - convert batches to block and write to the ledger
-func (c *Chain) processBatch( batch *msgs.QEntry) error{
+func (c *Chain) processBatch(batch *msgs.QEntry) error {
 	var envs []*common.Envelope
 	for _, requestAck := range batch.Requests {
-		reqMsg,err := c.fetchRequest(requestAck)
+		reqMsg, err := c.fetchRequest(requestAck)
 		if err != nil {
 			return errors.WithMessage(err, "Cannot fetch request from Request Store")
 		}
-		env,err:= protoutil.UnmarshalEnvelope(reqMsg.Data)
+		env, err := protoutil.UnmarshalEnvelope(reqMsg.Data)
 		if err != nil {
 			return errors.WithMessage(err, "Cannot Unmarshal Request Data")
 		}
 		if c.isConfig(env) {
-			configBlock := c.CreateBlock([]*common.Envelope{env})
-			c.writeConfigBlock(configBlock)
+			reconfig, err := c.getNewReconfiguration(env)
+
+			if err != nil {
+				return errors.Errorf("Cannot Generate Reconfiguration: %s", err)
+			}
+			c.Node.PendingReconfigurations = append(c.Node.PendingReconfigurations, reconfig)
+			//JIRA FLY2-106 set config envelope
+			err = c.setConfigEnv(reconfig[1].GetNewConfig(), env)
```

Comment:
I think the getters and setters of the hash map can be removed - instead, use a slice for the pending config envelopes as is done for `PendingReconfigurations` - which is appended with a config envelope at this point. When a `networkConfig` is passed to Snap by Mir it is always going to correspond to the least recent `PendingReconfiguration` and therefore the least recent config envelope. This relation could be made more idiomatic by creating a slice whose elements are a struct with fields for the pending reconfiguration and it's accompanying config envelope.

Furthermore there are situations where the current hashmap fails as an ordering service node admin may send different configuration envelopes which result in the same Mir network state.

Reply:
@harrymknight ave some doubts regarding this.Wouldn't be hashmap is more effective in searching
In hashmap , we are pairing network config hash to the corresponding common envelope
if we use an array, need to search through the array or can I assume that only need pop the last element in the array?

Reply:
@harrymknight  code is fixed with review comments